LeViT v6: add LeViT model and update configs

tdarwish25 opened the pull request:
This PR introduces the new LeViT v6 model and updates supporting configs.

Added:
- workloads/LeViT/LeViT_v6.py (new model)

Updated:
- ttsim/config/simconfig.py
- ttsim/front/functional/op.py
- ttsim/ops/op.py
- ttsim/config/wl2archmap.py
- config/all_archs.yaml
- config/wl2archmapping.yaml
- config/ip_workloads.yaml

Summary:
- Adds LeViT v6 model implementation
- Updates simulation config and operator definitions
- Refreshes workload-to-architecture mapping
- Extends workload configurations to include LeViT v6

ramamalladiTT commented on ttsim/ops/op.py:
```diff
@@ -2373,13 +2096,10 @@ def get_perf_counts(self, inT, outT, **kwargs):
         if self.perf_stats is not None:
             return self.perf_stats
 
-        is_backprop = kwargs.get('is_backprop', False)
```
why is this being edited out?

ramamalladiTT commented on ttsim/ops/op.py:
```diff
@@ -2586,15 +2282,15 @@ def SimOpFactory(optype: str) -> type[SimOp]:
+            LeakyReluOp          : ['LeakyRelu'],
+            SigmoidOp            : ['Sigmoid'],
+            ResizeOp             : ['Resize'],
+            ReduceSumOp          : ['ReduceSum'],
```
ReduceSumOp exists in main. Redo the merge conflict resolution.

ramamalladiTT commented on workloads/LeViT/LeViT_v5.py:
Is this required given you have v6?

tdarwish25 replied:
Agreed. this will be removed.

ramamalladiTT commented on ttsim/config/simconfig.py:
```diff
@@ -19,6 +19,24 @@
 DEBUG = LOG.debug
 
 
+def canonical_op(name: str) -> str:
```
why we need it? We will have to call what is defined by the ONNX spec. HardSwish.

tdarwish25 replied:
The ttsim/config edits were incremental workarounds for LeViT to align with the customer baseline and fix some of the Hardswish errors during debug, not broader model changes

ramamalladiTT commented on ttsim/front/functional/op.py:
```diff
@@ -504,6 +484,15 @@ def UniversalOperator(name, /, optype, params, ipos, **kwargs):
+# Hardswish frontend symbols:
+# - Use 'Hardswish' to match existing workloads (e.g., LeViT_v5.py uses F.Hardswish)
+# - Keep 'HardSwish' as an alias for compatibility
+Hardswish     = partial(UnaryOperator, optype='HARDSWISH')
```
Use this instead is better:

```
HardSwish = partial(UnaryOperator, optype='HardSwish')
```

tdarwish25 replied:
I initially ran into errors while debugging Hardswish, which led me to temporarily rename it as HARDWISH in the YAML and config ops flags. I’ll revisit this in V6 with your proposed changes and validate them further to ensure proper handling.

tdarwish25 commented on ttsim/ops/op.py:
```diff
@@ -1033,44 +921,32 @@ def get_perf_counts(self, inT, outT, **kwargs):
         else:
-            # Traditional AveragePool with explicit kernel_shape
-            kernel_shape = self.attrs.get('kernel_shape')  # Required attribute
+            kernel_shape = self.attrs.get('kernel_shape')
```
These edits were part of my workaround for the Hardswish and ReduceSumOp during debugging. only addressed LeViT-related cleanup and uncalled files. I'll do further checking..

tdarwish25 replied:
Removed today to align with main.

tdarwish25 commented on workloads/LeViT/LeViT_v6.py:
Good point. Aligned! Once the conflicts are resolved and the proposed errors are fixed, I plan to update the naming LeViT.py

tdarwish25 replied:
Plan to rename LeViT_v6 to LeViT once no further critical changes are identified, making it the baseline workload

tdarwish25 commented:
This PR is now up to date with main and incorporates Rama’s latest feedback.

**Summary of changes:**

- Updated LeViT v6 (workloads/LeViT/LeViT_v6.py) with new changes and aligned configs.

- Removed LeViT v5 (deprecated and superseded by v6).

- Resolved YAML conflicts:
 
- ip_workloads.yaml → merged LeViT6 with UNet + llama2 contributions.
 
- wl2archmapping.yaml → integrated llama2 updates cleanly.
 
**Cleaned up unused code:**
 
- Removed get_default_wl2arch_typespec from ttsim/config/wl2archmap.py and LeViT v6.
 
- Removed def canonical_op() from ttsim/config/simconfig.py.
 
**Conflict in ttsim/ops/op.py resolved by aligning with main and integrating only safe changes:**
 
- Added Hardswish to EltwiseUnaryOp (consistent naming).
 
- Kept ReduceSum mapping consistent with main and enabled forward support.
 
- No other logic changes; all existing operator implementations and perf accounting remain untouched.
 
- Standardized naming: replaced "HARDSWISH" with "Hardswish" across factory, YAML configs, and LeViT v6.
 
- Reverted YOLOv7 renames back to original names to avoid breaking YOLO workloads.

tdarwish25 commented on ttsim/ops/op.py:
```diff
@@ -1033,44 +921,32 @@ def get_perf_counts(self, inT, outT, **kwargs):
             input_height, input_width = input_shape[-2], input_shape[-1]
             output_height, output_width = output_size
 
-            # Calculate kernel size, stride, and padding to achieve the desired output size
```
Reverted the change back to mainline behavior in ttsim/ops/op.py.

tdarwish25 commented on ttsim/ops/op.py:
```diff
@@ -2389,9 +2109,11 @@ def get_perf_counts(self, inT, outT, **kwargs):
-
-        print(np_out)
+            axis = self.attrs.get('axis', None)
+            if axis is None:
```
Rebased ttsim/ops/op.py to main and reapplied only the minimal LeViT-specific changes to avoid drift.

tdarwish25 commented on config/all_archs.yaml:
```diff
@@ -15,6 +15,7 @@ ipblocks:
             -   {name: sub, tpt: { int8: 2048, bf16: 512, fp16: 512, fp8: 2048,}}
             -   {name: mul, tpt: { int8: 2048, bf16: 512, fp16: 512, fp8: 2048,}}
             -   {name: mac, tpt: { int8: 2048, bf16: 512, fp16: 512, fp8: 2048,}}
+            -   { name: HARDSWISH, tpt: { int8: 2048, bf16: 512, fp16: 512, fp8: 2048 } }  # added
```
Deleted.

tdarwish25 commented on config/ip_workloads.yaml:
```diff
@@ -4,7 +4,7 @@ workloads:
     name: YOLOv7
     basedir: workloads
-    module : YOLO7@YOLO_v7.py
+    module : YOLO7@Yolo_v7.py
```
Fixed YOLO file naming: reverted changes and aligned with main to restore baseline compatibility.

tdarwish25 commented on config/wl2archmapping.yaml:
```diff
@@ -39,10 +39,10 @@ op_fusion_spec:
 op_rsrc_spec:
   compute:
-    matrix: [Matmul, Conv]
+    matrix: [Matmul, Conv, HARDSWISH, ConvTranspose] #added HARDSWISH
```
Fixed and added to the vector.

tdarwish25 commented on ttsim/config/wl2archmap.py:
```diff
@@ -228,3 +228,19 @@ def get_wlmapspec_from_yaml(cfg_yaml_file) -> WL2ArchMap:
+
+
+# New: simple default typespec for direct-script runs and tests
+def get_default_wl2arch_typespec(global_type: str = "fp32",
```
Done.

tdarwish25 commented:
PR ready for re-review

Conflicts resolved and feedback incorporated (except final rename: LeViT_v6 → LeViT, to be done after approval).
Scope limited to LeViT only; shared files restored to match main.
Hardswish added as SimOp and mapped under vector compute; no HW instruction added.
Removed duplicate top-level files (ip_workloads.yaml, wl2archmapping.yaml, op.py).
Validation:
Global dryrun passes.
LeViT v6 runs pass via CLI and Python.
Sanity checks on other models pass; Polaris is stable with no regressions observed.
Note: After approval, I will rename LeViT_v6.py to LeViT.py in a follow-up commit. Thanks!

ramamalladiTT commented on config/all_archs.yaml:
```diff
@@ -32,7 +32,8 @@ ipblocks:
             -   {name: mov,   tpt: {int8: 128.,  int32: 32., fp32: 32. }}
             -   {name: max,   tpt: {int8: 128.,  int32: 32., fp32: 32. }}
-
+           
```
revert this change.

ramamalladiTT commented on ttsim/front/functional/op.py:
```diff
@@ -271,26 +271,6 @@ def __call__(self, *xargs):
         return self.otensor
 
-#sequence a list of layers...
-# run as a pipeline:
```
do not delete these comments.

tdarwish25 commented:
**Update Summary:**

- Synced branch fea/levit_v5 with main.

- Added two updated files:

               ttsim/front/functional/op.py – includes WL2ArchTypeSpec compatibility fix and Hardswish operator support.

               config/all_archs.yaml – reverted to align with main branch version.

All changes verified locally and tested under the latest Polaris CLI run for both LeViT and Yolo7 as sanity check.

Branch is now clean and up to date for review.

tdarwish25 commented:
The branch has been rebased and squashed into a single commit on top of main, and all updates have been validated locally for sanity check with the only following files changes:
-   workloads/LeViT/LeViT.py 
-   ttsim/ops/op.py 
-   ttsim/front/functional/op.py 
-   config/ip_workloads.yaml 
-   config/wl2archmapping.yam
